src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -3,91 +3,276 @@
-            Checksum solutionChecksum,
-            ReadOnlyMemory<Checksum> checksums,
-            CancellationToken cancellationToken)
+        // Create a channel to communicate between the searching and writing tasks.  This allows the searching task to
+        // find items, add them to the channel synchronously, and immediately continue searching for more items.
+        // Concurrently, the writing task can read from the channel and write the items to the pipe-writer.
+        var channel = Channel.CreateUnbounded<(Checksum checksum, object asset)>(new UnboundedChannelOptions()
         {
-            using var writer = new ObjectWriter(stream, leaveOpen: true, cancellationToken);
+            // We have a single task reading the data from the channel and writing it to the pipe.  This option
+            // allows the channel to operate in a more efficient manner knowing it won't have to sychronize data
+            // for multiple readers.
+            SingleReader = true,
+
+            // Currently we only have a single writer writing to the channel when we call FindAllAssetsAsync.
+            // However, we could change this in the future to allow the search to happen in parallel.
+            SingleWriter = true,
+        });
+
+        // When cancellation happens, attempt to close the channel.  That will unblock the task writing the assets
+        // to the pipe.
+        using var _ = cancellationToken.Register(() => channel.Writer.TryComplete(new OperationCanceledException(cancellationToken)));
+
+        // Keep track of how many checksums we found.  We must find all the checksums we were asked to find.
+        var foundChecksumCount = 0;
+
+        // Spin up a task to go search for all the requested checksums, adding results to the channel.
+        var findAssetsTask = FindAllAssetsAsync();
+
+        // Spin up a task to read from the channel and write out the assets to the pipe-writer.
+        var writeAssetsTask = WriteAllAssetsToPipeAsync();
 
-            // This information is not actually needed on the receiving end.  However, we still send it so that the
-            // receiver can assert that both sides are talking about the same solution snapshot and no weird invariant
-            // breaks have occurred.
-            solutionChecksum.WriteTo(writer);
+        // Wait for both the searching and writing tasks to finish.
+        await Task.WhenAll(findAssetsTask, writeAssetsTask).ConfigureAwait(false);
+        cancellationToken.ThrowIfCancellationRequested();
 
-            // special case
-            if (checksums.Length == 0)
-                return;
+        // If we weren't canceled, we better have found and written out all the expected assets.
+        Contract.ThrowIfTrue(foundChecksumCount != checksums.Length);
 
-            Debug.Assert(assetMap != null);
+        return;
 
-            foreach (var (checksum, asset) in assetMap)
+        async Task FindAllAssetsAsync()
```

Comment:
>async [](http://example.com/codeflow?start=8&length=5)

nit: this is a me thing, but I find it difficult to follow where the locals of the containing method are all being used/updated when there are this many nested functions. Any thoughts on making at least some of these functions static and passing across the necessary state?

Reply:
Upgrade your brain!

Reply:
changed to statics.